packages/itwinui-react/src/core/SearchBox/SearchBox.tsx

```diff
@@ -0,0 +1,139 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import cx from 'classnames';
+import { InputProps } from '../Input';
+import {
+  Icon,
+  InputFlexContainer,
+  useTheme,
+  SvgSearch,
+  SvgCloseSmall,
+} from '../utils';
+import { IconButton } from '../Buttons';
+type SearchBoxOwnProps =
+  | {
+      expandable?: false;
+      isExpanded?: undefined;
+      onToggle?: undefined;
+      animateTo?: undefined;
+      searchIcon?: React.ReactNode;
+    }
+  | {
+      /**
+       * Whether the searchbox is animated to expand.
+       * @default false
+       */
+      expandable: true;
+      /**
+       * Animation direction.
+       * @default 'right'
+       */
+      animateTo?: 'left' | 'right';
+      /**
+       * Whether or not to show the search input.
+       * @default false
+       */
+      isExpanded?: boolean;
+      /**
+       * Callback function for toggling an expansion state.
+       */
+      onToggle?: (isExpanding: boolean) => void;
+      /**
+       * Initial icon for searchbox.
+       * default: `search` when searchbox is closed, `close` when searchbox is expanded.
+       */
+      searchIcon?: React.ReactNode;
+    };
+
+export type SearchBoxProps = SearchBoxOwnProps & InputProps;
+
+const defaultSearchIcon = (isExpanded: boolean) => {
+  return isExpanded ? <SvgCloseSmall /> : <SvgSearch />;
+};
+
+/**
+ * Searchbox component.
+ * Used for searches :)
+ * @example
+ *  <SearchBox>
+ *    <input type='search' placeholder='Search...' />
+ *  </SearchBox>
+ *
+ *  <SearchBox>
+ *    <input type='search' placeholder='Search...' />
+ *    <IconButton styleType='borderless'>
+ *     <SvgCaretUpSmall />
+ *   </IconButton>
+ *   <IconButton styleType='borderless'>
+ *      <SvgCaretDownSmall />
+ *     </IconButton>
+ *  </SearchBox>
+ */
+export const SearchBox = (props: SearchBoxProps) => {
+  useTheme();
+
+  const [localIsExpanded, setLocalIsExpanded] = React.useState(false);
+
+  const {
+    size,
+    expandable = false,
+    animateTo = 'right',
+    isExpanded = localIsExpanded,
+    onToggle = setLocalIsExpanded,
+    searchIcon = defaultSearchIcon(isExpanded),
+    children,
+    ...rest
+  } = props;
+
+  const animatedVersion = () => {
+    return (
+      <>
+        {animateTo === 'right' && (
+          <IconButton
+            styleType='borderless'
+            onClick={() => onToggle(!isExpanded)}
```

Comment:
You don't need javascript for this. You can do it fully in CSS if you only use `<input>` (no `<button>`) and then expand on focus (using `:focus-within`).

Try replacing contents of App.tsx in the vite playground with this:
<details>

<summary>App.tsx</summary>

```jsx
import styled from '@emotion/styled';

const App = () => {
  return (
    <>
      <InputWrapper>
        <input />
      </InputWrapper>
    </>
  );
};

const InputWrapper = styled.div`
  display: inline-flex;
  border: 1px solid;
  background: Canvas;
  color: CanvasText;
  cursor: pointer;
  position: relative;
  height: 2.5rem;
  width: 3rem;
  transition: width 0.2s;

  > input {
    all: unset;
    width: 100%;
    padding-inline: 4px;
  }

  &::before {
    content: '';
    position: absolute;
    inset: 0;
    background-color: inherit;
    pointer-events: none;
  }

  &::after {
    --search-svg: url('data:image/svg+xml,<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">\
      <path d="M11 9.7c.7-1 1.1-2.2 1.1-3.5.1-3.5-2.7-6.2-6-6.2C2.7 0 0 2.7 0 6.1s2.7 6.1 6.1 6.1c1.3 0 2.5-.4 3.5-1.1l4.9 4.9 1.4-1.4zm-5 .5c-2.3 0-4.1-1.8-4.1-4.1S3.7 2 6 2s4.1 1.8 4.1 4.1-1.8 4.1-4.1 4.1"/>\
    </svg>');

    content: '';
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    background-color: currentColor;
    mask: var(--search-svg) no-repeat center / contain;
    width: 1rem;
    height: 1rem;
    pointer-events: none;
  }

  &:focus-within {
    cursor: unset;
    width: 100%;
    border-color: hotpink;

    &::before,
    &::after {
      content: unset;
    }
  }
`;

export default App;

```
</details>

Reply:
You can add a data attribute in addition to `:focus-within`. That will make it possible to make it user controlled.

For icon, I think it's actually fine to not support custom icons. It is possible to achieve through CSS variables, but not very important for initial featureset imo.